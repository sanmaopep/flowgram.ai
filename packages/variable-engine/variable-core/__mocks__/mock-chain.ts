import { ScopeChain, Scope } from '../src';

/**
 * 规则：
 * - Global 覆盖所有 Scope，所有 Scope 依赖 Global
 * - 存在环路：cycle1 -> cycle2 -> cycle3 -> cycle1
 */

export class MockScopeChain extends ScopeChain {
  getDeps(scope: Scope): Scope[] {
    const res: Scope[] = [];
    if (scope.id === 'global') {
      return [];
    }

    const global = this.variableEngine.getScopeById('global');
    if (global) {
      res.push(global);
    }

    // 模拟循环依赖场景
    if (String(scope.id).startsWith('cycle')) {
      return this.variableEngine
        .getAllScopes()
        .filter((_scope) => String(_scope.id).startsWith('cycle') && _scope.id !== scope.id);
    }

    return res;
  }

  getCovers(scope: Scope): Scope[] {
    if (scope.id === 'global') {
      return this.variableEngine.getAllScopes().filter(_scope => _scope.id !== 'global');
    }

    // 模拟循环依赖场景
    if (String(scope.id).startsWith('cycle')) {
      return this.variableEngine
        .getAllScopes()
        .filter((_scope) => String(_scope.id).startsWith('cycle') && _scope.id !== scope.id);
    }

    return [];
  }

  sortAll(): Scope[] {
    return this.variableEngine.getAllScopes();
  }
}
