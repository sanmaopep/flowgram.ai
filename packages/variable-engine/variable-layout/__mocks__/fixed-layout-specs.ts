import { FlowDocumentJSON } from '@flowgram.ai/document';

export const fixLayout1: FlowDocumentJSON = {
  nodes: [
    {
      type: 'start',
      meta: {
        isStart: true,
      },
      id: 'start',
    },
    {
      type: 'getRecords',
      id: 'getRecords_07e97c55832',
    },
    {
      type: 'loop',
      id: 'forEach_260a8f85ff2',
      blocks: [
        {
          type: 'createRecord',
          id: 'createRecord_8f85ff2c11d',
        },
        {
          type: 'dynamicSplit',
          id: 'exclusiveSplit_ff2c11d0fb4',
          blocks: [
            {
              id: 'branch_f2c11d0fb42',
            },
            {
              id: 'branch_2c11d0fb42c',
            },
          ],
        },
      ],
    },
    {
      type: 'dynamicSplit',
      id: 'exclusiveSplit_88dbf2c60ae',
      meta: {
        defaultExpanded: true,
      },
      blocks: [
        {
          id: 'branch_8dbf2c60aee',
          blocks: [
            {
              type: 'dynamicSplit',
              id: 'exclusiveSplit_a59afaadc9a',
              blocks: [
                {
                  id: 'branch_59afaadc9ac',
                },
                {
                  id: 'branch_9afaadc9acd',
                  blocks: [
                    {
                      type: 'deleteRecords',
                      id: 'deleteRecords_c32807e97c5',
                    },
                  ]
                },
              ],
            },
          ]
        },
        {
          id: 'branch_dbf2c60aee4',
          blocks: [
            {
              type: 'updateRecords',
              id: 'updateRecords_7ed2a172c32',
            },
          ]
        },
      ],
    },
    {
      type: 'end',
      id: 'end',
    },
  ],
};
