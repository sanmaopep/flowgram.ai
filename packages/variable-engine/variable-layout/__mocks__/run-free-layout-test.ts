import { describe, expect, test } from 'vitest';
import { VariableEngine } from '@flowgram.ai/variable-core';
import { ASTKind } from '@flowgram.ai/variable-core';
import {
  FlowNodeVariableData
} from '../src';
import { TestConfig, getContainer } from './container';
import { WorkflowDocument, WorkflowJSON } from '@flowgram.ai/free-layout-core';

export const runFreeLayoutTest = (testName: string, spec: WorkflowJSON, config?: TestConfig) => {
  describe(testName, async () => {
    const container = getContainer('free', config);
    const flowDocument = container.get(WorkflowDocument);
    const variableEngine = container.get(VariableEngine);

    variableEngine.onScopeChange(action => {
      const { type, scope } = action;

      // 作用域默认创建一个默认变量
      if (type === 'add') {
        scope.ast.set('/', {
          kind: ASTKind.VariableDeclaration,
          type: ASTKind.String,
          key: String(scope.id),
        });
      }
    });

    // 创建一个全局作用域
    variableEngine.createScope('testScope');

    const traverseVariableDatas = () =>
      flowDocument
        .getAllNodes()
        // TODO 包含 $ 的节点不注册 variableData
        .filter(_node => !_node.id.startsWith('$'))
        .map(_node => _node.getData(FlowNodeVariableData))
        .filter(Boolean);

    // 初始化所有节点的私有作用域
    const initAllNodePrivate = () => {
      traverseVariableDatas().forEach(_data => {
        _data.initPrivate();
      });
    };

    // 初始化所有节点的可用变量
    const printAllNodeAvailableMapping = (_scopeType: 'public' | 'private' = 'public') =>
      traverseVariableDatas().reduce((acm, _data) => {
        const scope = _data[_scopeType]!;
        acm.set(String(scope.id), scope.available.variableKeys);

        return acm;
      }, new Map<string, string[]>());

    // 初始化所有节点的覆盖作用域
    const printAllCovers = (_scopeType: 'public' | 'private' = 'public') =>
      traverseVariableDatas().reduce((acm, _data) => {
        const scope = _data[_scopeType]!;
        acm.set(
          String(scope.id),
          scope.coverScopes.map(_scope => String(_scope.id)),
        );

        return acm;
      }, new Map<string, string[]>());

    await flowDocument.fromJSON(spec);

    test('test get Deps', () => {
      expect(printAllNodeAvailableMapping()).toMatchSnapshot();
    });

    test('test get Covers', () => {
      expect(printAllCovers()).toMatchSnapshot();
    });

    test('test get Deps After Init Private', () => {
      initAllNodePrivate();
      expect(printAllNodeAvailableMapping()).toMatchSnapshot();
    });

    test('test get private scope Deps', () => {
      expect(printAllNodeAvailableMapping('private')).toMatchSnapshot();
    });

    test('test get Covers After Init Private', () => {
      expect(printAllCovers()).toMatchSnapshot();
    });

    test('test get private scope Covers', () => {
      expect(printAllCovers('private')).toMatchSnapshot();
    });

    test('test sort', () => {
      expect(variableEngine.getAllScopes({ sort: true }).map(_scope => _scope.id)).toMatchSnapshot();
    });

    config?.runExtraTest?.(container);
  });


}
