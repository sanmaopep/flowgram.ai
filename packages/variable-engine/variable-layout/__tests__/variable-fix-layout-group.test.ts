import { runFixedLayoutTest } from '../__mocks__/run-fixed-layout-test';

runFixedLayoutTest(
  'Variable Fix Layout Group',
  {
    nodes: [
      {
        id: 'start_0',
        type: 'start',
        meta: {
          isStart: true,
        },
        blocks: [],
      },
      {
        id: '$group_test$',
        type: 'block',
        blocks: [
          {
            id: 'node_0',
            type: 'noop',
            blocks: [],
          },
          {
            id: 'node_1',
            type: 'noop',
            blocks: [],
          },
        ],
      },
      {
        id: 'end_0',
        type: 'end',
        blocks: [],
      },
    ],
  },
  {}
);
