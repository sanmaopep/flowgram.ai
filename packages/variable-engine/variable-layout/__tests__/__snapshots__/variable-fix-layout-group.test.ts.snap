// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Variable Fix Layout Group > test get Covers 1`] = `
Map {
  "start_0" => [
    "node_0",
    "node_1",
    "end_0",
  ],
  "node_0" => [
    "node_1",
    "end_0",
  ],
  "node_1" => [
    "end_0",
  ],
  "end_0" => [],
}
`;

exports[`Variable Fix Layout Group > test get Covers After Init Private 1`] = `
Map {
  "start_0" => [
    "node_0",
    "node_0_private",
    "node_1",
    "node_1_private",
    "end_0",
    "end_0_private",
  ],
  "node_0" => [
    "node_1",
    "node_1_private",
    "end_0",
    "end_0_private",
  ],
  "node_1" => [
    "end_0",
    "end_0_private",
  ],
  "end_0" => [],
}
`;

exports[`Variable Fix Layout Group > test get Deps 1`] = `
Map {
  "start_0" => [],
  "node_0" => [
    "start_0",
  ],
  "node_1" => [
    "start_0",
    "node_0",
  ],
  "end_0" => [
    "start_0",
    "node_0",
    "node_1",
  ],
}
`;

exports[`Variable Fix Layout Group > test get Deps After Init Private 1`] = `
Map {
  "start_0" => [
    "start_0_private",
  ],
  "node_0" => [
    "start_0",
    "node_0_private",
  ],
  "node_1" => [
    "start_0",
    "node_0",
    "node_1_private",
  ],
  "end_0" => [
    "start_0",
    "node_0",
    "node_1",
    "end_0_private",
  ],
}
`;

exports[`Variable Fix Layout Group > test get private scope Covers 1`] = `
Map {
  "start_0_private" => [
    "start_0",
  ],
  "node_0_private" => [
    "node_0",
  ],
  "node_1_private" => [
    "node_1",
  ],
  "end_0_private" => [
    "end_0",
  ],
}
`;

exports[`Variable Fix Layout Group > test get private scope Deps 1`] = `
Map {
  "start_0_private" => [],
  "node_0_private" => [
    "start_0",
  ],
  "node_1_private" => [
    "start_0",
    "node_0",
  ],
  "end_0_private" => [
    "start_0",
    "node_0",
    "node_1",
  ],
}
`;

exports[`Variable Fix Layout Group > test sort 1`] = `
[
  "start_0_private",
  "start_0",
  "node_0",
  "node_0_private",
  "node_1",
  "node_1_private",
  "end_0",
  "end_0_private",
  "globalScope",
  "$group_test$",
]
`;
